docs/testing/writing-unit-tests.md

```diff
@@ -0,0 +1,456 @@
+
+### Set up and tear down
+Most tests that do anything useful will need some setup, some code execution, and possibly some tear down / resetting of resources that are common to all specs. This allows you to minimize code repetition, and is done through use of `beforeEach/All` and `afterEach/All`. These run before or after each `it` function gets run. Here's an example:
+
+```Javascript
+define([
+    'myModule',
+    'common/runtime'
+], (MyModule, Runtime) => {
+    describe('Test MyModule', () => {
+        let runtime;
+        beforeEach(() => {
+            // set up the Runtime object that manages various state elements
+            runtime = Runtime.make();
+        });
+
+        afterEach(() => {
+            // The runtime also puts itself in the global space, so this
+            // tears it back down
+            window.kbaseRuntime = null;
+        });
+
+        it('should do things that need runtime', () => {
+            let bus = runtime.bus();
+            expect(MyModule.doThingsWithBus(bus)).toBeTruthy();
+        });
+    });
+})
+```
+The above sets up and takes down the Runtime module, gets a message bus from it, and uses it for a test spec. Something more realistic would use the same Runtime in multiple specs. Note that `beforeEach` and `afterEach` (as well as `it`) can take `async` functions. More on that later.
+
+This also makes use of Javascript closure. The `runtime` variable is declared in the outer scope, it's set in `beforeEach`, then when each spec is called, it's ready to go.
+
+### Asynchrony
+
+The nature of modern Javascript is asynchronous code. Most of the interesting things will be interacting with either KBase services, or with the user, and so we don't want those to run synchronously and locking up the browser while processing.
+
+Jasmine is capable of testing asynchronous code as well. There's 3 idioms that can be used.
+
+***1. Async/Await***
+
+If you're testing code that returns a Promise or uses the async/await idiom, this can be done as follows:
+
+```Javascript
+it('tests async functions', async () => {
+    const result = await runAsyncFunction();
+    expect(result).toBeTruthy();
+});
+```
+
+Easy enough, just treat it as a common Promise function. This should work with any KBase code that returns a Promise.
+
+***2. Return the Promise***
+
+Another option when working with Promises or similar *thenable* code, is to just return the Promise from the spec and embed the `expect` function in a then block. Something like this:
+
+```Javascript
+it('returns a promise', () => {
+    return codeThatMakesAPromise()
+        .then((result) => {
+            expect(result).toBeDefined();
+        });
+});
+```
+This is similar to the above, but it can also be useful for trapping errors that you want to see (in a `.catch` block instead of `.then`). Though that can also be done with `async/await` in a `try/catch`. 
```

Comment:
Do you want to add in an example for this, e.g. using jasmine's `fail(...)`? Alternatively, point towards the Jasmine docs on running async tests (https://jasmine.github.io/tutorials/async)?